### PR TITLE
Added element mixin for surface quadrature

### PR DESCRIPTION
**Pull request by WillTrojak**
This allows new point sets to be generated for plugins, including surface quadratures.

### CODE REVIEW

File: pyfr/plugins/base.py [hunk 1]
```diff
@@ -167,3 +169,90 @@ def __init__(self, intg, *args, **kwargs):
...
+
+        # Apply the operator to the mesh elements and reshape
+        e_surf = eles.eles[:, eidx]
+        ploc = plocop @ e_surf.reshape(nspts, -1)
+        ploc = ploc.reshape(len(pts), neles, ndims)
+
+        return ploc
+
+    @staticmethod
+    def jacs_at(eles, eidx, pts):
```
Comment (FreddieWitherden) [inline pyfr/plugins/base.py:220]: There looks to be a lot of code duplication here.  It should be possible to eliminate most of this by extending some of the eles `_at_np` methods to support more flexible `name` arguments (see `rcpdjac_at_np` for an example).

File: pyfr/plugins/base.py [hunk 2]
```diff
@@ -167,3 +169,90 @@ def __init__(self, intg, *args, **kwargs):
...
+
+            q = get_quadrule(itype, qrule, qdeg=qdeg, flags='s')
+            ppts, pwts = self._proj_pts(proj, q.pts), q.wts
+            self._surf_quad[itype, proj] = (ppts, pwts)
+
+        return ppts, pwts
+
+    @staticmethod
+    def _proj_pts(projector, pts):
+        pts = np.atleast_2d(pts.T)
```
Comment (FreddieWitherden) [inline pyfr/plugins/base.py:202]: Is this check needed in this context?

File: doc/src/user_guide.rst [hunk 3]
```diff
@@ -1253,12 +1253,22 @@ vectors to a CSV file. Parameterised with
 
     ``(x, y, [z])``
 
+5. ``quad-deg-{etype}`` --- degree of quadrature rule for fluidforce
```
Comment (FreddieWitherden) [inline doc/src/user_guide.rst:1256]: force

File: pyfr/plugins/base.py (hunk 2 above)
Comment (FreddieWitherden) [reply pyfr/plugins/base.py:202]: Just following up here.

File: pyfr/plugins/fluidforce.py [hunk 4]
```diff
@@ -105,17 +113,11 @@ def __init__(self, intg, cfgsect, suffix):
...
-
-                eidxs[etype, fidx].append(eidx)
-                norms[etype, fidx].append(pnorms)
-
                 # Get the flux points position of the given face and element
                 # indices relative to the moment origin
                 if self._mcomp:
-                    fpts_idx = eles.basis.facefpts[fidx]
-                    rfpt = eles.plocfpts[fpts_idx, eidx] - morigin
+                    ploc = eles.plocs_at(ppts)[:,[eidx]]
```
Comment (FreddieWitherden) [inline pyfr/plugins/fluidforce.py:119]: See if we can go without the [].  Also, spacing,

File: pyfr/solvers/base/elements.py [hunk 5]
```diff
@@ -277,24 +278,28 @@ def qpts(self):
...
-        if np.any(np.sqrt(mag_pnorm_fpts) < 1e-10):
+        if np.any(np.sqrt(mag_pnorm) < 1e-10):
             raise RuntimeError('Zero face normals detected')
 
-        return pnorm_fpts
+        return pnorm
 
     @cached_property
-    def _smats_djacs_mpts(self):
+    def smats_djacs_mpts(self):
```
Comment (FreddieWitherden) [inline pyfr/solvers/base/elements.py:302]: Why is this renaming needed?

File: pyfr/plugins/base.py [hunk 6]
```diff
@@ -167,3 +169,35 @@ def __init__(self, intg, *args, **kwargs):
...
                 self._ele_region_data[f'{etype}_idxs'] = eidxs
+
+
+class SurfaceMixin:
+    def __init__(self, intg, *args, **kwargs):
+        super().__init__(intg, *args, **kwargs)
+
+        self._surf_quad = {}
+
+    def surf_quad(self, itype, proj, *args):
```
Comment (FreddieWitherden) [inline pyfr/plugins/base.py:180]: Just make this a quasi-static method and avoid the need for a new member variable `_surf_quad`.

Comment (FreddieWitherden) [reply pyfr/plugins/base.py:180]: Or you can @memoize this (but remove *args).

Comment (WillTrojak) [reply pyfr/plugins/base.py:180]: I added the `*args` so you could just pass `eles.basis.faces[fidx]` but I think this is a niche use case.

File: pyfr/solvers/base/elements.py (hunk 5 above)
Comment (WillTrojak) [reply pyfr/solvers/base/elements.py:302]: It shouldn't be needed anymore.

File: pyfr/plugins/fluidforce.py (hunk 4 above)
Comment (WillTrojak) [reply pyfr/plugins/fluidforce.py:119]: Sadly not, I tried initially as this is ugly.

File: pyfr/plugins/base.py (hunk 2 above)
Comment (WillTrojak) [reply pyfr/plugins/base.py:202]: I think it's a reasonable check in case someone tries to pass 1D quad points.

File: pyfr/plugins/base.py (hunk 6 above)
Comment (FreddieWitherden) [reply pyfr/plugins/base.py]: Yeah, it is usually just asking for trouble.

File: pyfr/plugins/base.py (hunk 2 above)
Comment (FreddieWitherden) [reply pyfr/plugins/base.py:204]: It is not really a check.  It is to add support for 1D projections, which I do not think make sense here (we want to get an exception).

File: pyfr/plugins/base.py [hunk 7]
```diff
@@ -167,3 +170,36 @@ def __init__(self, intg, *args, **kwargs):
 
             if not isinstance(eidxs, slice):
                 self._ele_region_data[f'{etype}_idxs'] = eidxs
+
+
+class SurfaceMixin:
+    def __init__(self, intg, *args, **kwargs):
```
Comment (FreddieWitherden) [inline pyfr/plugins/base.py:176]: Do we need this?

File: pyfr/plugins/base.py [hunk 8]
```diff
@@ -167,3 +170,36 @@ def __init__(self, intg, *args, **kwargs):
...
+            # Quadrature rule (default to that of the solution points)
+            qrule = self.cfg.get(self.cfgsect, f'quad-pts-{itype}', rname)
+            try:
+                qdeg = self.cfg.getint(self.cfgsect, f'quad-deg-{itype}')
+            except NoOptionError:
+                qdeg = self.cfg.getint(self.cfgsect, 'quad-deg')
+
+            q = get_quadrule(itype, qrule, qdeg=qdeg, flags='s')
+            ppts, pwts = self._proj_pts(proj, q.pts), q.wts
+            self._surf_quad[itype, proj] = (ppts, pwts)
```
Comment (FreddieWitherden) [inline pyfr/plugins/base.py:198]: Do we need this?

File: pyfr/solvers/base/elements.py [hunk 9]
```diff
@@ -277,21 +278,25 @@ def qpts(self):
...
-        smats = self.smat_at_np('fpts').transpose(1, 3, 0, 2)
+        return self.pnorm_at('fpts', self.basis.norm_fpts)
+
+    @memoize
+    def pnorm_at(self, name, norm):
+        smats = self.smat_at_np(name).transpose(1, 3, 0, 2)
 
         # We need to compute |J|*[(J^{-1})^{T}.N] where J is the
-        # Jacobian and N is the normal for each fpt.  Using
+        # Jacobian and N is the normal for each pnt. Using
```
Comment (FreddieWitherden) [inline pyfr/solvers/base/elements.py:288]: point?

File: pyfr/plugins/base.py [hunk 10]
```diff
@@ -167,3 +170,26 @@ def __init__(self, intg, *args, **kwargs):
 
             if not isinstance(eidxs, slice):
                 self._ele_region_data[f'{etype}_idxs'] = eidxs
+
+
+class SurfaceMixin:
+    @memoize
+    def surf_quad(self, itype, proj):
```
Comment (FreddieWitherden) [inline pyfr/plugins/base.py:177]: Make _ as probably not meant to be called by external users of the plugin?

File: pyfr/plugins/fluidforce.py [hunk 11]
```diff
@@ -86,12 +86,20 @@ def __init__(self, intg, cfgsect, suffix):
 
             for etype, eidx, fidx, flags in mesh[bc].astype('U4,i4,i1,i2'):
                 eles = elemap[etype]
+                itype, proj, norm = eles.basis.faces[fidx]
 
-                if (etype, fidx) not in m0:
-                    facefpts = eles.basis.facefpts[fidx]
+                ppts, pwts = self.surf_quad(itype, proj)
```
Comment (FreddieWitherden) [inline pyfr/plugins/fluidforce.py:91]: q is probably a better prefix as they're quadrature points.

File: pyfr/plugins/base.py [hunk 12]
```diff
@@ -167,3 +170,26 @@ def __init__(self, intg, *args, **kwargs):
...
+        rname = self.cfg.get(f'solver-interfaces-{itype}', 'flux-pts')
+
+        # Quadrature rule (default to that of the solution points)
+        qrule = self.cfg.get(self.cfgsect, f'quad-pts-{itype}', rname)
+        try:
+            qdeg = self.cfg.getint(self.cfgsect, f'quad-deg-{itype}')
+        except NoOptionError:
+            qdeg = self.cfg.getint(self.cfgsect, 'quad-deg')
+
+        q = get_quadrule(itype, qrule, qdeg=qdeg, flags='s')
```
Comment (FreddieWitherden) [inline pyfr/plugins/base.py:188]: Make flags an option argument to surf_quad?

**Pull request merged**